### PR TITLE
Detect if FDBLock is stolen by another writer

### PR DESCRIPTION
**Pull request by rnewson**
Ensure that a crashed JVM won't leave directories locked forever.

### CODE REVIEW

**Comment by rnewson** [comment]
an FDBLock can be safely stolen by another writer. The new writer will delete the lock and recreate it with its own unique (random) value. The old lock is atomically invalid. The next ensureValid check will fail and the associated IndexWriter will automatically close. IndexWriter always wraps the directory with a LockValidatingDirectoryWrapper which calls lock.ensureValid before all destructive methods (notably, createOutput and rename).

File: src/main/java/com/cloudant/fdblucene/FDBLock.java [hunk 1]
```diff
@@ -17,19 +17,39 @@
...
     private final Directory dir;
+    private final UUID uuid;
     private final String name;
+    private boolean invalid = false;
     private boolean closed = false;
 
-    FDBLock(final Directory dir, final String name) {
+    public static Lock obtain(final Directory dir, final UUID uuid, final String name) throws IOException {
+        try {
+            try (final IndexOutput output = dir.createOutput(name, null)) {
```
Comment (tonysun83) [inline src/main/java/com/cloudant/fdblucene/FDBLock.java]: this nested try with out catch looks weird

**Comment by tonysun83** [comment]
I'm not seeing where you use wallclock/timestamp to generate the uuid? Seems you're using the directory's random uuid for comparison?

File: src/main/java/com/cloudant/fdblucene/FDBLock.java (hunk 1 above)
Comment (rnewson) [reply src/main/java/com/cloudant/fdblucene/FDBLock.java]: that is the java 8 AutoCloseable thing. It means output.close() will be called when scope ends.

**Comment by rnewson** [comment]
the uuid is a random 128 bit value, we can assume duplicates don't happen.

The lock is checked for validity just before a destructive operation, and the act of stealing the lock makes all other locks invalid.

**Pull request merged**